compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -2203,13 +2203,11 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
         // will end up taking too much memory. If it does, we should just limit
         // how much GADT constraints we infer - it's always sound to infer less.
         pat1.putAttachment(InferredGadtConstraints, ctx.gadt)
-      if (pt1.isValueType) // insert a cast if body does not conform to expected type if we disregard gadt bounds
+      if pt1.isValueType then // insert a cast if body does not conform to expected type if we disregard gadt bounds
         body1 = body1.ensureConforms(pt1)(using originalCtx)
-      val nni = pat1.notNullInfo.seq(
-        guard1.notNullInfoIf(false).alt(
-          guard1.notNullInfoIf(true).seq(body1.notNullInfo)
-        )
-      )
+      val nni = pat1.notNullInfo
+        .seq(guard1.notNullInfoIf(false).alt(guard1.notNullInfoIf(true)))
```

Comment:
If the guard is false, the body does not execute. I don't think it's sound to .seq the body after an .alt that includes the case of the guard being false.

Reply:
I'm still working on the fix for typedCase:

```scala
def test3(i: Int) =
  var x: String | Null = null
  i match
    case 1 if x != null => ()
    case _ => x = " "
  x.trim() // ok
```